Quaver.Shared/IPC/QuaverIpcHandler.cs

```diff
@@ -54,7 +62,78 @@ public static void HandleProtocolMessage(string message)
+            }
+            else if (message.StartsWith("map/"))
+            {
+                var game = (QuaverGame) GameBase.Game;
+                if (game.CurrentScreen.Type != QuaverScreenType.Select)
+                {
+                    NotificationManager.Show(NotificationLevel.Warning,
+                        "You must be in the song select screen to select a map!");
+                    return;
+                }
+
+                var screen = game.CurrentScreen as SelectionScreen;
+
+                message = message.Replace("map/", "");
+                int mapId;
+                if (!int.TryParse(message, out mapId))
+                {
+                    NotificationManager.Show(NotificationLevel.Warning, "The provided ID was not a number!");
+                    return;
+                }
+
+                // Check if we have the map installed
+                var map = MapManager.FindMapFromOnlineId(mapId);
+                if (map != null)
+                {
+                    MapManager.Selected.Value = map;
+
+                    lock (screen.AvailableMapsets.Value)
+                        screen.AvailableMapsets.Value = MapsetHelper.FilterMapsets(screen.CurrentSearchQuery);
+                    return;
+                }
+
+                if (!OnlineManager.Connected)
+                {
+                    NotificationManager.Show(NotificationLevel.Warning, "You must be logged in to download maps!");
+                    return;
+                }
+
+                // Check if the map exists online
+                var request = new APIRequestMapInformation(mapId);
+                var response = request.ExecuteRequest();
+                if (response.Status == 400 || response.Map?.MapsetId == -1)
+                {
+                    NotificationManager.Show(NotificationLevel.Warning, "The map does not exist!");
+                    return;
+                } else if (response.Status != 200)
+                {
+                    NotificationManager.Show(NotificationLevel.Error, "Something happened during the request");
+                    return;
+                }
+
+                // User doesn't have the map, so download it for them
+                if (MapsetDownloadManager.CurrentDownloads.All(x => x.MapsetId != response.Map.MapsetId))
+                {
+                    var download = MapsetDownloadManager.Download(response.Map.MapsetId, response.Map.Artist, response.Map.Title);
+                    MapsetDownloadManager.OpenOnlineHub();
+
+                    // Auto import
+                    download.Completed.ValueChanged += (sender, args) =>
+                    {
+                        if (game.CurrentScreen.Type == QuaverScreenType.Select)
+                        {
+                            var selectScreen = (SelectionScreen) game.CurrentScreen;
+                            game.CurrentScreen.Exit(() => new ImportingScreen(null, true));
```

Comment:
Is there a way to select the correct map after the set has imported? So you don't have to handle the message twice.

Reply:
Import already automatically selects the last imported map

Reply:
Mapset or map?

Reply:
Right... fixed now